### ISSUE TITLE
Fine grained notifications

### ISSUE THREAD

**Issue by bmunkholm** [realm/realm-js#296]

**Comment by edmofro** [comment]
Any eta on this? Would be very helpful :)

**Comment by alazier** [comment]
No eta yet but we are currently working on this.

**Comment by appden** [comment]
Here's a summary of my thoughts on the API we should have for this. We should continue with the trend of using a subset of the Node [`EventEmitter`](https://nodejs.org/api/events.html#events_class_eventemitter) API, since it's familiar to JS devs and used in many libraries.

#### `Realm.Results`

Adding listeners to a `Results` instance should retain it until all listeners have been removed, **or** when the Realm has been closed.

```js
let results = realm.objects('Person');

results.addListener('change', (event) => {
  // This closure implicitly retains `results`, but we should
  // *explicitly* retain it until all listeners are removed.

  // It's idiomatic for callbacks to just have one "event" parameter.
  // All properties should be read-only getters below...
  event.realm;  // nice-to-have convenience
  event.target;  // target of the event, `results` in this case
  event.type;  // "change" in this case
  event.deletions;
  event.insertions;
  event.modifications;
});

// This should *not* remove the listener since it's a different results instance.
realm.objects('Person').removeAllListeners();
```

#### `Realm.List`

Adding listeners to a `List` instance should likewise retain it, but there are more considerations here. When getting a `List` instance from a Realm object in JS, we expose them as different JS objects even though they really access the same underlying data. So listeners shouldn't rely on an individual JS `List` instance.

We should use the fact that `core` "interns" `LinkView` objects to make sure we store all listeners in one place. I personally think the easiest thing to do here would be to use `LinkView` as a key to store alive `List` instances for each JS context, and return back the same JS `List` instance when retrieved. That way we'd also get this wonderful behavior: `object.listProp === object.listProp`

```js
realmObject.listProperty.addListener('change', (event) => {
  // same event object as described above
});

// This *should* actually remove the listener above.
realmObject.listProperty.removeAllListeners();
```

#### `Realm.Object`

This is a more tricky, but similar case to `List` above. Since we don't have unique identifiers for objects in Realm, it would be less performant to expose uniqued JS Realm objects for each Realm object accessor, though it **is** possible by using `std::map` instead of `std::unordered_map` to store them per JS context and `Table`.

So I think the compromise should be to continue expose different JS objects for the same underlying Realm object, but store listeners in a map by their underlying Realm object (i.e. `Table` and `Row`). Obviously, listeners should be removed when Realm objects are deleted or the Realm is closed.

```js
list[0].addListener('change', (event) => {
  event.realm;  // nice-to-have convenience
  event.target;  // target of the event, `list[0]` object in this case
  event.type;  // "change" in this case
  event.changes;  // calling it "modifications" like above might be confusing
});

// This *should* work remove the listener above!
list[0].removeAllListeners();
```

### Additional Notes

Right now the listener callbacks for `Realm` instances receive **two** parameters: Realm instance and event type. We should definitely change this to be an `event` like those above with read-only `realm` and `type` properties.

Like `EventEmitter`, **many** JS listener APIs also have `on()` as a synonym for `addListener()`. It would be trivial for us to do the same, and so we should. The `once()` method from `EventEmitter` is also really easy to add and super appreciated by developers when the need arises. I'd really like for us to have that as well.

The end. :smile:

**Comment by astigsen** [comment]
+1 nice writeup

I have also gotten feedback about the missing `on()` alias.

**Comment by edmofro** [comment]
Agree, very nice writeup.

Was hoping as part of this change the current realm level notifications would pass the object type and object itself to the listeners. This would be especially helpful for our project which syncs records out to a central server when they change; adding a global listener that is given the record directly seems like the easiest way to ensure all database changes are detected and the appropriate record synced.

**Comment by alazier** [comment]
@edmofro I think the use case you describe is something we may want to support - you could build what you are asking for on top of the `Results` notifier described by @appden but it would be easier to simply have a new notification type you can register directly on a `Realm` object that provides objects of a given type which have been added, removed, or changed.

**Comment by alazier** [comment]
@appden - I know in the past we have talked about more detailed notification types beyond `change`. If I recall correctly, you proposed something like `change:<objectType>` in which case you would get change events when any objects of type `objectType` are added, removed, or changed. I think if we supported this it would address what @edmofro is requesting. Perhaps this could also be used to support listening for changes to individual properties - we could support something like `change:<objectType>:<propertyName>` for this use case. I think this would fit well with the current design - the only thing I think we are missing from this design which is available in kvo is the ability to request the old values for changed properties.

One other big question we need to address is when these notifications will fire. In the cocoa binding notifications on Results and List are async, while other notifications are synchronous (I think this is correct). I'm not convinced this is what we want to be doing for JS. In my limited experiments so far I have already run into issues trying to integrate async Results notifications with react-native apps so we may want to reconsider how and when these notifications fire. What are your thoughts on this issue?

**Comment by alazier** [comment]
@bigfish24 Adding a notifier on `Results` will always cause things to be asynchronous rather than getting notifications when the write transaction completes as is the case with other notifications. Users have also asked for the actual objects which have been added/removed rather than indexes which makes more sense when dealing with unordered sets as opposed to `Results` where you need the indexes to update any UI properly.

**Comment by appden** [comment]
@alazier I had proposed `"change:<propertyName>"` as a possible solution for KVO on Realm objects, which is inspired by [Backbone's API](http://backbonejs.org/#Events-catalog). I actually think this feature should be outside the scope of addressing this issue, unless it's trivial to add during implementation.

I think we should just continue to have collections notifications be async since it would require some major architectural changes to make them sync, along with introducing serious performance issues. I think there are good arguments to be made that `change` events first async on Realms and Realm objects, though I think it's fine to just leave it the way it is now (which is sync for changes made in that thread, and async otherwise).

As for how we integrate async notifications with React Native, I don't think that'll be a problem at all, but I'd like to actually see the code to better understand what you issues ran into.

**Comment by alazier** [comment]
@appden have you actually tried using the current notifications as they are currently implemented? There are problems integrating with React Native as any case where you rely on notifications to update your ui there will be a time that there is a mismatch between what you are reading from the Realm and the most recent information you have from the notification system. In the experiments I have run this has caused things to behave and crash in unexpected ways. Try out the branch I made a while back and run the example if you want a more concrete example of this issue.

**Comment by bigfish24** [comment]
@alazier is this the same issue detailed in [realm-cocoa](https://github.com/realm/realm-cocoa/issues/3665) where you apply changes due to UI interactions immediately, but then a notification fires async?

**Comment by bigfish24** [comment]
@appden I agree KVO functionality should not be a priority, we should seek parity with realm-cocoa as first step and improve the capabilities of the notifications from there.

**Comment by alazier** [comment]
@bigfish24 we should not seek parity with cocoa if the design of the cocoa notification system does not work for JS. I think the issue is similar to what you are referencing - if we use the notification system as is then you can not access a `Results` on which notifications are registered outside of the notification block without inadvertently re-runing the the query synchronously and killing the notification. This would happen in any case you want to show the `Results` object in the UI which I would say is a pretty common user scenario.

**Comment by alazier** [comment]
And as far as key value observing goes, we should design our overall notification apis to take into account what we anticipate kvo will look like in the end even if we don't build it initially. It would not make sense to ship a set of apis that we know we will need to change at a later date to support kvo.

**Comment by astigsen** [comment]
> if we use the notification system as is then you can not access a Results on which notifications are registered outside of the notification block without inadvertently re-runing the the query synchronously and killing the notification.

On cocoa this can only happen if you access the `Results` right after doing a local write transaction (and it has to be one that changes the objects you are querying on). In that case it is forced to rerun the query synchronously since you would otherwise be looking at stale and possibly totally incorrect data.

Would this be different on js?

**Comment by appden** [comment]
@astigsen We'd still want that `change` event to fire. I believe we could just copy the `Results` under the hood (in C++) so those notifications are still delivered.

**Comment by astigsen** [comment]
> We'd still want that change event to fire

Of course, but you won't know if there actually was a change (which affected that specific `Results` happening until after you have re-run the query.

Why would you need to copy the `Results`?

**Comment by appden** [comment]
I believe we'd need to make a copy because the `ResultsNotifier` will not send out a notification if the query has been re-run on that `Results` instance synchronously.

**Comment by astigsen** [comment]
Ok, but if that is a problem, it seems better to fix that behaviour rather than copying the `Results`.

You might also want to keep an eye on https://github.com/realm/realm-cocoa/pull/3439 which deals with a somewhat related problem.

**Comment by appden** [comment]
Ok, I had thought that behavior was intended, but @alazier will take care of it. :smile:

**Comment by tgoyne** [comment]
IMO JS needs async writes, not sync notifications.

The async notification mechanism somewhat fundamentally doesn't work if write transactions are done on the same thread as notifications are registered. Even if you're careful to access anything after your write transaction, unless you do something gross like display a full-screen spinner that blocks the UI, the user can always try to access data which you haven't already read in the window between the commit and when the notifications fire, and then you're in an inconsistent state.

However, sync notifications aren't really a solution for this. You could fake sync notifications with the current design by just blocking the main thread until the notifications are ready to deliver, but this is obviously terrible. Notifications designed to work synchronously wouldn't be much better, though. They'd be a little faster, but in a lot of realistic cases the overhead of the asyncness is pretty trivial compared to that of just running the query and calculating the changeset. Blocking the UI for potentially hundreds of ms isn't acceptable, and I don't think it makes sense to spend time on a solution which inherently leads to that.

Async writes solve the problem entirely, and are something we presumably want anyway. A perhaps simpler shorter-term solution would be to use a second SharedGroup for write transactions with implicit handover from the read one to the write one (note that this doesn't actually work at the moment), which would let us leave the read transaction at the pre-write state until the notification is ready. Users would probably be confused by this, though.

**Comment by alazier** [comment]
@tgoyne thanks for the suggestions. If we make notifications fire only after an async write callback, I can see how for many cases we can avoid accessing a `Results` object and inadvertently running a query synchronously.

There are two primary use cases where this still wouldn't help though. The first is if you were to access the `Results` object during the write transaction - for this case we can advise users to simply not do this or to use multiple `Results` objects to prevent this from happening. 

The more problematic scenario is when other components are using a `Results` instance to render UI. This is the more common scenario and we nor the user have control over how 3rd party components are using the data that is being passed in. It seems to me the only way to reliably support this scenario would be to support deep snapshots, in which case the pattern we would promote would involve passing in a new snapshot and other supporting data whenever a change notification is received. Conceptually this should work reliably with or without async writes. Is if this is something we can provide in the short term without major changes to core? We could just keep around old SharedGroups for snapshotted objects but with the risk of running into the file size issue. If we could find a way to mitigate the risk or somehow detect and prevent unreasonable file size expansion I think this would be worth exploring. Any ideas?

**Comment by astigsen** [comment]
> The more problematic scenario is when other components are using a Results instance to render UI. 

A component can be in three possible states in relation to it's data:

1. It is up-to-date. The `Results` matches what it is drawing on the screen. - Great!

2. There has just been a write transaction, but the `Results` has not finished updating yet, and no notification has been sent yet (since we won't know if it has changed until after the query has been re-run). - BAD, a re-draw at this time would at best force a synchronous refresh of the query, possibly making the UI stutter. Could also break some invariants within the component, since it has not been notified of the change.

3. `Results` has been refreshed and a notification has been sent to the component. - Great! It can now update any invariants and redraw the UI at will.

The best solution to avoid state number two seems to be making writes async, which should be the most "node like" behaviour anyways. If writes are async the component should only ever be possible to be in state one or three, no snapshots needed.

**Comment by alazier** [comment]
I don't see how async writes reliably prevent scenario 2, especially in the case that we don't own the component the `Results` is being passed to. There is nothing to prevent an event from firing between the async write and the notification block which could prematurely access the `Results`.

**Comment by astigsen** [comment]
The point is that the async write works on its own version of the Realm, and the original `Results` (and the Realm it belongs to) won't be updated until the async write has been completed and the queries has been re-run, at which point it will send the notifications.

So any event firing in the interim will catch the component in either state one or three.

**Comment by alazier** [comment]
So if I am understanding correctly, you are proposing that writes not be visible to other consumers or until all accessors, queries, and changesets have been updated and made ready for consumption? I can see how this would prevent scenario 2. There is still the situation where you could access the updated Results before the notification has fired. This is not as bad as re-running the query synchronously but it could still result in unexpected behavior. Async writes also do not address the scenario where a consumer needs access to a stable version of a Realm over multiple events, between which changes may be made. Snapshotting seems to be the only viable solution to that problem, although we can deal with this outside the scope of notifications.

**Comment by appden** [comment]
I think async writes could be a nice addition for performance reasons, among other things, but I definitely don't think they should be necessary to move forward with `change` notifications. We can have `change` notifications with sync writes, and we can also continue to have queries computed synchronously when `Results` are accessed. We can encourage best practices for the ideal performance, like only accessing `Results` within `change` callbacks, but I think our APIs should generally continue to work the way they do now.

I agree with @alazier that deep snapshots would make some things easier on the application developer, and would make running async queries make more sense, but like he said, they are outside the scope of notifications.

**Comment by cmelchior** [comment]
Just a note about this since Java just went through a similar discussion.

The conclusion we came to is that there is no proper way of mixing synchronous writes and async queries if you want to to have a consistent behaviour and/or trigger changelisteners right after a commit.

For that reason we came to the conclusion that the only safe way is to only trigger changelisteners on the next run-loop event. This makes it completely safe as we then have one point in time where we can coordinate running async queries/writes and determining when we should trigger notifications.

So we have done 2 things:

1) We now log a warning to the log if you mix asynchronous queries and local commits. This might later throw an actual exception.

2) All changelisteners are only triggered on the next run-loop. No exceptions.
